examples/with-firebase-authentication-serverless/pages/_app.js

```diff
@@ -0,0 +1,88 @@
+/* eslint react/jsx-props-no-spreading: 0 */
+/* globals window */
+import React from 'react'
+import PropTypes from 'prop-types'
+import { get, set } from 'lodash/object'
+import { AuthUserInfoContext, useFirebaseAuth } from '../utils/auth/hooks'
+import { createAuthUser, createAuthUserInfo } from '../utils/auth/user'
+
+const App = props => {
+  const { AuthUserInfo, Component, pageProps } = props
+
+  // We'll use the authed user from client-side auth (Firebase JS SDK)
+  // when available. On the server side, we'll use the authed user from
+  // the session. This allows us to server-render while also using Firebase's
+  // client-side auth functionality.
+  const { user: firebaseUser } = useFirebaseAuth()
+  const AuthUserFromClient = createAuthUser(firebaseUser)
+  const { AuthUser: AuthUserFromSession, token } = AuthUserInfo
+  const AuthUser = AuthUserFromClient || AuthUserFromSession || null
+  return (
+    <AuthUserInfoContext.Provider value={{ AuthUser, token }}>
+      <Component {...pageProps} />
+    </AuthUserInfoContext.Provider>
+  )
+}
+
+App.getInitialProps = async ({ Component, ctx }) => {
```

Comment:
I'd rather not disable static optimization for this entire application. Can we avoid this?

Reply:
Yes. I moved the logic in `_app.js` to a higher-order component, so getting the authenticated user is optional for each page. The small downside is that statically-rendered pages won't include the authenticated user as JSON in the HTML, so navigating from a static page to a page that requires the authenticated user will render *without* the authenticated user until the Firebase client JS gets the user. For example, in this demo load `/example`, then click to "home". This isn't a big deal, because people can handle it with loading states—just wanted to note it here.